gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -219,11 +245,35 @@ parse_single_string_literal (AST::DelimTokenTree &invoc_token_tree,
   else if (parser.peek_current_token ()->get_id () == last_token_id)
     rust_error_at (invoc_locus, "macro takes 1 argument");
   else
-    rust_error_at (invoc_locus, "argument must be a string literal");
+    {
+      bool has_error = false;
+
+      auto expanded_expr
+	= try_expand_many_expr (parser, last_token_id, expander, has_error);
```

Comment:
This function tries and parse multiple expressions from an invocation. Think something like `concat!(a, b, 15, "hey")`. This function would parse and return a vector containing `a`, `b`, `15` and `"hey"`. Here, we need to make sure we're only parsing one expression, which could be a literal expression or a macro invocation. So you need to add some more checks after calling this:
```suggestion
      auto expanded_exprs
	= try_expand_many_expr (parser, last_token_id, expander, has_error);
      if (expanded_exprs.size() > 1)
          rust_error_at(invoc_locus, "expected only one argument, got %lu", expanded_exprs.size());
          
      // then check for literal expr or macro invocation etc
```

but I think it would be even simpler to not call into `try_expand_many_expr`. If you look above in the function, we call `parser.parse_literal_expr()`. You could add a call to `parser.parse_macro_invocation({})` and see which one returns a null pointer.